src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+// - detect if acquiring partitions takes too long
+// - replace Task.Delay with something more sophisticated
+
+namespace Microsoft.Health.Events.EventHubProcessor
+{
+    public class PartitionLockingBackgroundService : BackgroundService
+    {
+        private string _processorId;
+
+        private List<int> _ownedPartitions;
+
+        private BlobContainerClient _containerClient;
+
+        private int _totalProcessorsRunning = 0;
+
+        private string[] _eventHubPartitions;
+
+        private EventBatchingService _eventBatchingService;
+
+        private StorageCheckpointClient _checkpointClient;
+
+        private ITelemetryLogger _logger;
+
+        private PartitionLockingBackgroundServiceOptions _partitionLockingOptions;
+
+        private EventHubClientOptions _eventHubClientOptions;
+
+        public PartitionLockingBackgroundService(
+            PartitionLockingBackgroundServiceOptions partitionLockingOptions,
+            EventHubClientOptions eventHubClientOptions,
+            EventBatchingService eventBatchingService,
+            StorageCheckpointClient checkpointClient,
+            ITelemetryLogger logger)
+        {
+            _processorId = Guid.NewGuid().ToString(); // consider using the pod name when running in AKS
+            _ownedPartitions = new List<int>();
+            _partitionLockingOptions = partitionLockingOptions;
+            _eventHubClientOptions = eventHubClientOptions;
+            _eventBatchingService = eventBatchingService;
+            _checkpointClient = checkpointClient;
+            _logger = logger;
+        }
+
+        private void CreateContainerClient()
+        {
+            var tokenCredential = EnsureArg.IsNotNull(_partitionLockingOptions.StorageTokenCredential, nameof(_partitionLockingOptions.StorageTokenCredential));
+            _containerClient = new BlobContainerClient(_partitionLockingOptions.BlobContainerUri, tokenCredential.GetCredential());
+            _containerClient.CreateIfNotExists();
+        }
+
+        private async Task RegisterProcessor(CancellationToken ct)
+        {
+            // write process id to storage
+            _logger.LogTrace($"Processor Id: {_processorId}");
+            BlobClient blobClient = _containerClient.GetBlobClient($"activepods/{_processorId}");
+
+            var activeUntil = new Dictionary<string, string>()
+            {
+                { "activeuntil", DateTime.UtcNow.AddMinutes(1).ToString() },
+            };
+
+            BlobUploadOptions options = new BlobUploadOptions() { Metadata = activeUntil };
+            await blobClient.UploadAsync(BinaryData.FromString(string.Empty), options, ct);
+
+            _logger.LogTrace($"Registered processor {_processorId}");
+        }
+
+        private async Task RegisterProcessorLoop(CancellationToken ct)
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                await RegisterProcessor(ct);
```

Comment:
Why do we need to run this loop? It looks like `_processorId` never changes, so why do we need to keep registering it every 15 seconds?

Reply:
we use the blob time to determine if a processor is active or not. If the pod is still emitting the processor id to blob and/or the blob is less than 1 minute old then we consider the processor to still be active. So we have a background task that periodically uploads the processor id to blob so that we can compute how many processors are currently active